### PR TITLE
[NO SQUASH] Inventory: Fix order of callbacks when swapping items

### LINKED ISSUE #13486: Incorrect sequence of calling callbacks `on_take` & `on_put` on swapping items in inventory
##### Minetest version
```
5.7.0
```

##### OS / Hardware
Operating system: Kubuntu 22.04.2 LTS

##### Summary
When swapping items in inventory, MT calls:
 - at first `on_put`
 - and then `on_take`
 
 It will be better if MT calls it in reverse order:
  - `on_take`
  - `on_put`

##### Steps to reproduce
For example, with LOTT game:
1. Install [LOTT](https://content.minetest.net/packages/Amaz/lordofthetest/)
2. * _also you can add debug prints into log [here](https://github.com/minetest-LOTR/Lord-of-the-Test/blob/32eab389274593f91f7f1625d6846bc5f43143e7/mods/lottarmor/armor.lua#L424)_
3. Open inventory, creative tab, search&take "Mithril boots" and "Training boots"
4. Open Main tab and put on Mithril one
5. Take ~wooden~ Training one and swap it with Mithril

As result:
 - you have no boots on preview in inventory
 - you have no boots on player model (F7)
 - and also you have no protection
image

### LINKED ISSUE #14210: unexpected behavior related to stack size in register_on_player_inventory_action callback
### Minetest version

```true
Minetest 5.9.0-dev-55f40a7f8 (Linux)
Using Irrlicht 1.9.0mt13
Using LuaJIT 2.1.1699392533
BUILD_TYPE=Release
RUN_IN_PLACE=1
USE_CURL=1
USE_GETTEXT=1
USE_SOUND=1
STATIC_SHAREDIR="."
STATIC_LOCALEDIR="locale"
```

### Active renderer

_No response_

### Irrlicht device

_No response_

### Operating system and version

Linux nail 6.6.1-1-MANJARO #1 SMP PREEMPT_DYNAMIC Thu Nov  9 04:27:49 UTC 2023 x86_64 GNU/Linux

### CPU model

_No response_

### GPU model

_No response_

### OpenGL version

_No response_

### Summary

@Niklp09 noticed that blocks went missing when shift-clicking shaped blocks using the invsaw of my [moreblocks/stairsplus fork](https://github.com/fluxionary/minetest-moreblocks). i've finally been able to reproduce exactly why that happens, which is because either the "count" value passed with the info to `minetest.register_on_player_inventory_action` is wrong, or i'm not understanding how to interpret the that value correctly. https://github.com/minetest/minetest/blob/995c19287468a5733956fdba028f940901d4d1d5/doc/lua_api.md?plain=1#L5713

### Steps to reproduce

download and use the following mod: https://github.com/fluxionary/minetest-invbug. it has no requirements.

in game, run the `/invbug` command. shift-click the stack of 33 items. then shift-click the stack of 99 items. you will see the following output in chat:

```
took 33 (from_stack="" to_stack="invbug:item 33")
took 99 (from_stack="invbug:item 33" to_stack="invbug:item 99")
took 33 (from_stack="invbug:item 66" to_stack="invbug:item 33")
```

both bottom entries are generated by the second shift-click. the first indicates that the stack of 33 in the first slot now has 99 items in it, and the second indicates that a new stack of 33 was created. the last line in particular makes little sense to me.

if these numbers are correct, can someone explain how to properly calculate the number of items that were moved (should be 132 in total).

### PR DESCRIPTION
**Pull request by SmallJoker**
This PR yet again touches the item movement code to fix the wrong order of put/take callbacks when swapping items.

Current swap callback order:
 * destination inventory put
 * source inventory take
 * source inventory put
 * destination inventory take

Corrected/expected callback order:
 * destination + source inventory take
 * source + destination inventory put

Revealed/discovered by @fluxionary in https://github.com/minetest-mods/unified_inventory/issues/208

Fixes #13486, fixes #14210

## To do

This PR is Ready for Review.

## How to test

1. MTG creative inventory: must work as expected
2. devtest chest: put and take any number of items, shift-click them. ensure the reported counts are correct.
    * Namely, bug #10805 must not re-appear.
3. unified_inventory : tab "Bags" -> put a bag in, replace with another. Open the bag and there must be no warning about a missing inventory list.
4. Modify `helper_moveaction.lua` L2/L3: change `on_take` to `on_whatever` and check for correctly failing unittests

**How to test (variant 2)**

1. Testing script: https://github.com/SmallJoker/minetest_lab/blob/master/init_76_inv_callbacks.lua
2. MTG creative mode enabled (no additional inventory mod)
3. Place `default:chest_locked` nearby
4. `/fs` and move items

Expected order: "allow take"+"allow put", then "on take" and finally "on put". Example:
```
	2                              | action=take (singleplayer) | [main,2] default:glass 4
	on_metadata_inventory_take     | [main,3] butterflies:butterfly_red 2 (singleplayer)
2023-06-12 21:06:01: ACTION[Server]: singleplayer takes butterflies:butterfly_red from chest at (-9,2,1)
	on_metadata_inventory_put      | [main,3] default:glass 4 (singleplayer)
2023-06-12 21:06:01: ACTION[Server]: singleplayer moves default:glass to chest at (-9,2,1)
	2                              | action=put (singleplayer) | [main,2] butterflies:butterfly_red 2
```

### CODE REVIEW

**Comment by sfan5** [comment]
@savilli I know you have a PR for this that hasn't been merged yet (sorry), do you have anything to comment here? does your PR include this fix?

**Comment by savilli** [comment]
No, my code has another order: src tale, src put, dst take, dst put.
Why is this order correct?

**Comment by kilbith** [comment]
Has it been _**extensively**_ tested with different inventories?

**Comment by SmallJoker** [comment]
If you'd like to ensure that it fulfills your definition of ***extensively***, feel free to use script N° 76 on the devtest chest (might need adjustments) and player inventory.

**Comment by fluxionary** [comment]
> @savilli A good example is a turntable to exchange items in real life. You first put the goods, then take it on both sides once the transaction is complete. This is also what I expect from item swaps in Minetest.

this is also the model i personally prefer. without some sort of true transaction system, it is the most fair, and the pattern i use in my smartshop fork:

https://github.com/fluxionary/minetest-smartshop/blob/58fbd0b1c7202c52c7ca235bb50b32d59bc54483/api/purchase_mechanics.lua#L122-L136

you both put the goods, then you both take the goods, and any errors are recorded (or corrected) afterwards. 

there's clearly race conditions lurking, but without some sort of actual synchronization API, it's better than the alternatives.

**Comment by savilli** [comment]
Well, I don't think this PR should be a bugfix, because it's not a bug. The callback order was just undefined.
If you want to define it this way, it's good for me. Just mention it in the docs, to make it clear that the order is established and shouldn't be changed.
Also, I think it would be nice to merge my PR #11885 and write a unittest that checks the order.

**Comment by SmallJoker** [comment]
#11885 was merged, so I extended the unittests to ensure the following:

1. non-empty ItemStacks in the inventory callbacks (indicates error)
2. correct execution of allow -> take -> put callbacks (according to new documentation)
3. proper functionality of the script interface to make use of the `inv` parameter in Lua

File: doc/lua_api.txt [hunk 1]
```diff
@@ -3300,6 +3300,18 @@ Player Inventory lists
...
+--------------------------
+
+This list assumes non-empty ItemStacks in both slots (swapping).
+Put/take callbacks about empty ItemStack are not executed.
+
+1. The "allow take" and "allow put" callbacks are each run once for the source
+   and destination inventory.
+2. The allowed ItemStacks are exchanged.
+3. The "on take" callback is run for source and destination inventory
+4. The "on put" callback is run for source and destination inventory
```
Comment (SmallJoker) [inline doc/lua_api.txt:3313]: I don't like this wording. Feel free to propose something better.

File: src/server/serverinventorymgr.h [hunk 2]
```diff
@@ -36,19 +36,19 @@ class ServerInventoryManager : public InventoryManager
 		m_env = env;
 	}
 
-	Inventory *getInventory(const InventoryLocation &loc);
```
Comment (savilli) [inline src/server/serverinventorymgr.h:39]: Are these changes necessary? What's wrong with their implementations in `ServerInventoryManager`?

File: src/script/cpp_api/s_base.h [hunk 3]
```diff
@@ -124,6 +124,7 @@ class ScriptApiBase : protected LuaHelper {
 	friend class ModApiBase;
 	friend class ModApiEnvMod;
 	friend class LuaVoxelManip;
+	friend class TestMoveAction; // needs getStack()
```
Comment (savilli) [inline src/script/cpp_api/s_base.h:127]: Maybe add `check_function` to `cpp_api` instead? Making all tests friend classes looks bad to me.

File: doc/lua_api.txt [hunk 4]
```diff
@@ -3300,6 +3300,18 @@ Player Inventory lists
     * Is not created automatically, use `InvRef:set_size`
     * Is only used to enhance the empty hand's tool capabilities
 
+ItemStack transation order
+--------------------------
+
+This list assumes non-empty ItemStacks in both slots (swapping).
+Put/take callbacks about empty ItemStack are not executed.
+
+1. The "allow take" and "allow put" callbacks are each run once for the source
```
Comment (savilli) [inline doc/lua_api.txt:3309]: I would split it into two lines.
1. The "allow take" callbacks are run for the source and destination inventory.
2. The "allow put" callbacks are run for the source and destination inventory.
And if we want to fix the order between the source and destination inventory: "run for the source and then for the destination inventory".

File: src/script/cpp_api/s_base.h (hunk 3 above)
Comment (SmallJoker) [reply src/script/cpp_api/s_base.h:127]: Yes, it is a bit hacky, however I would prefer to not integrate unittest code into the regular code base. The current approach reduces code complexity and hence maintenance. Furthermore, `check_function` is pretty trivial but might change for other unittests (different amount of parameters or return values).

File: src/server/serverinventorymgr.h (hunk 2 above)
Comment (SmallJoker) [reply src/server/serverinventorymgr.h:39]: We need make sure that the correct `getInventory` implementation is called, hence `virtual`. Furthermore, the scripting interface uses the server's `ServerInventoryManager`-type variable to retrieve inventories.

I will undo the other few function changes, though. They are currently not needed by the unittest anyway.

File: doc/lua_api.txt (hunk 4 above)
Comment (SmallJoker) [reply doc/lua_api.txt:3309]: Actually the `allow_take`/`allow_put` order does not matter because no items are actually moved. It is simply to determine them movable item count.

**Review by TurkeyMcMac — COMMENTED** [review]
The code looks good and works. There are just two notes I have.

File: src/inventory.cpp [hunk 5]
```diff
@@ -705,34 +705,32 @@ void InventoryList::moveItemSomewhere(u32 i, InventoryList *dest, u32 count)
...
 	if (this == dest && i == dest_i)
-		return count;
+		return moved;
 
 	// Take item from source list
-	ItemStack item1;
-	if (count == 0)
-		item1 = changeItem(i, ItemStack());
-	else
-		item1 = takeItem(i, count);
```
[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/inventory.cpp:719]: Occurrence 1: `IMoveAction::apply`. Value 0 is not possible due to previous guards.
Occurrence 2: ` IMoveAction::clientApply` Value 0 is not possible because empty itemstacks cannot be moved in the GUI.

[the remark this answers is not in the document]
Comment (TurkeyMcMac) [reply src/inventory.cpp:719]: Could you update this comment (remove "0 for everything") and remove the default value of the `count` argument? https://github.com/minetest/minetest/blob/84e5866f03974360f9a73df1023c855eadefe936/src/inventory.h#L268-L272

[the remark this answers is not in the document]
Comment (SmallJoker) [reply src/inventory.cpp:719]: :+1: Done.

**Review by TurkeyMcMac — APPROVED** [review]

**Comment by sfan5** [comment]
does this conflict with #11882?

**Comment by TurkeyMcMac** [comment]
Probably.

**Comment by sfan5** [comment]
I'd like to merge the other PR first since it also fixes some other bug. Thoughts?

**Comment by SmallJoker** [comment]
This PR conflicts with the other in a few places due to partially shared code changes. The other was waiting for unittests to ensure it's working correctly. I could try to implement this fix after that PR if that's preferred.

**Comment by Zughy** [comment]
@SmallJoker please rebase docs (see #13438)

**Comment by alek13** [comment]
can be merged ?

**Comment by SmallJoker** [comment]
The rebase took quite long but eventually I got it to work again. Retested with 
```
minetest --run-unittests --test-module TestMoveAction
```

File: builtin/game/tests/test_moveaction.lua [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+		-- testMovePartial
+		if info.stack:get_name() == "default:takeput_max_5" then
+			return 5
+		end
+
+		-- testCallbacks
+		if info.stack:get_name():find("default:takeput_cb_%d") then
+			if #exec_listing == 0 then
+				if inv:get_stack(info.listname, info.index):is_empty() then
```
Comment (sfan5) [inline builtin/game/tests/test_moveaction.lua:65]: shouldn't the c++ test code decide which execution pattern it wants to see?
(why not pass that into `__helper_check_callbacks`)

File: doc/lua_api.md [hunk 7]
```diff
@@ -3618,6 +3618,18 @@ Player Inventory lists
     * Is not created automatically, use `InvRef:set_size`
     * Is only used to enhance the empty hand's tool capabilities
 
+ItemStack transation order
+--------------------------
+
+This list assumes non-empty ItemStacks in both slots (swapping).
+Put/take callbacks about empty ItemStack are not executed.
```
Comment (sfan5) [inline doc/lua_api.md:3625]:
```suggestion
ItemStack transaction order
---------------------------

This list describes the situation for non-empty ItemStacks in both slots (swap operation).
Put/take callbacks on empty ItemStack are not executed.
```

File: src/inventory.h [hunk 8]
```diff
@@ -276,9 +276,9 @@ class InventoryList
 	ItemStack takeItem(u32 i, u32 takecount);
 
 	// Move an item to a different list (or a different stack in the same list)
-	// count is the maximum number of items to move (0 for everything)
-	// returns number of moved items
-	u32 moveItem(u32 i, InventoryList *dest, u32 dest_i,
+	// count is the maximum number of items to move
```
Comment (sfan5) [inline src/inventory.h:279]:
```suggestion
	// count is the maximum number of items to move (0 for everything)
```

File: src/inventory.cpp [hunk 9]
```diff
@@ -756,54 +756,52 @@ void InventoryList::moveItemSomewhere(u32 i, InventoryList *dest, u32 count)
...
 	// Try to add the item to destination list
-	count = item1.count;
-	item1 = dest->addItem(dest_i, item1);
+	ItemStack leftover = dest->addItem(dest_i, moved);
 
 	// If something is returned, the item was not fully added
-	if (!item1.empty()) {
-		bool nothing_added = (item1.count == count);
+	if (!leftover.empty()) {
+		moved.remove(leftover.count);
```
Comment (sfan5) [inline src/inventory.cpp:782]:
```suggestion
		// keep track of how many we actually moved
		moved.remove(leftover.count);
```

File: src/inventorymanager.cpp [hunk 10]
```diff
@@ -440,25 +442,20 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
...
-	count = src_item.count;
-	if (src_can_take_count != -1 && count > src_can_take_count)
-		count = src_can_take_count;
-	if (dst_can_put_count != -1 && count > dst_can_put_count)
-		count = dst_can_put_count;
+	// Apply limits given by allow_* callbacks
+	if (src_can_take_count != -1 && (int)move_count > src_can_take_count)
+		move_count = src_can_take_count;
+	if (dst_can_put_count != -1 && (int)move_count > dst_can_put_count)
+		move_count = dst_can_put_count;
```
Comment (sfan5) [inline src/inventorymanager.cpp:451]: clearer IMO
```suggestion
	if (src_can_take_count != -1)
		move_count = core::min_<u32>(src_can_take_count, move_count);
	if (dst_can_put_count != -1)
		move_count = core::min_<u32>(dst_can_put_count, move_count);
```

File: src/inventorymanager.cpp [hunk 11]
```diff
@@ -440,25 +442,20 @@ void IMoveAction::apply(InventoryManager *mgr, ServerActiveObject *player, IGame
...
 
-	/* Limit according to source item count */
-	if (count > list_from->getItem(from_i).count)
-		count = list_from->getItem(from_i).count;
+	// Limit just in case
+	if (move_count > list_from->getItem(from_i).count)
+		move_count = list_from->getItem(from_i).count;
```
Comment (sfan5) [inline src/inventorymanager.cpp:455]: not 100% sure I'd rather convert this to an assert
(should not happen at all looking at the rest of the code)

File: src/unittest/test_moveaction.cpp [hunk 12]
```diff
@@ -39,50 +40,39 @@ class TestMoveAction : public TestBase
...
-	ofs << helper_lua_src;
-	ofs.close();
-
 	ServerScripting server_scripting(&server);
 	try {
-		server_scripting.loadMod(Server::getBuiltinLuaPath() + DIR_DELIM "init.lua", BUILTIN_MOD_NAME);
-		server_scripting.loadMod(helper_lua, BUILTIN_MOD_NAME);
+		// FIXME: When removing the line below, the unittest does NOT crash
+		// (but it should) when running all unittests in order or registration.
+		server.m_modmgr = std::make_unique<ServerModManager>(server.m_path_world);
```
Comment (sfan5) [inline src/unittest/test_moveaction.cpp:62]: what's the mod manager used for?

File: src/unittest/test_moveaction.cpp [hunk 13]
```diff
@@ -218,10 +215,59 @@ void TestMoveAction::testSwapToUnallowed(ServerActiveObject *obj, IGameDef *game
...
-	UASSERT(inv.p2.getList("main")->getItem(0).getItemString() == "default:water 60");
+	UASSERT(inv.p2.getList("main")->getItem(0).getItemString() == "default:takeput_deny 60");
+}
+
+static bool check_function(lua_State *L)
+{
+	bool ok = false;
+	int error_handler = PUSH_ERROR_HANDLER(L);
+
+	lua_getglobal(L, "minetest");
```
Comment (sfan5) [inline src/unittest/test_moveaction.cpp:231]:
```suggestion
	lua_getglobal(L, "core");
```

File: src/unittest/test_moveaction.cpp [hunk 14]
```diff
@@ -218,10 +215,59 @@ void TestMoveAction::testSwapToUnallowed(ServerActiveObject *obj, IGameDef *game
...
+	lua_getfield(L, -1, "__helper_check_callbacks");
+	int result = lua_pcall(L, 0, 1, error_handler);
+	if (result == 0)
+		ok = lua_toboolean(L, -1);
+	else
+		puts(lua_tostring(L, -1)); // Error msg
```
Comment (sfan5) [inline src/unittest/test_moveaction.cpp:237]:
```suggestion
		errorstream << lua_tostring(L, -1); // error msg
```

File: src/unittest/test_moveaction.cpp [hunk 15]
```diff
@@ -218,10 +215,59 @@ void TestMoveAction::testSwapToUnallowed(ServerActiveObject *obj, IGameDef *game
...
+
+	inv.p1.addList("main", 10)->addItem(0, parse_itemstack("default:takeput_cb_1 10"));
+	inv.p2.addList("main", 10);
+
+	apply_action("Move 10 player:p1 main 0 player:p2 main 1", &inv, obj, server);
+
+	// Expecting 4 callback executions. See Lua file for details.
+	UASSERT(check_function(server->getScriptIface()->getStack()));
+
+	delete server->m_inventory_mgr.release();
```
Comment (sfan5) [inline src/unittest/test_moveaction.cpp:256]: `server->m_inventory_mgr.reset();`

**Review by sfan5 — COMMENTED** [review]
(code review only)

File: builtin/game/tests/test_moveaction.lua (hunk 6 above)
Comment (SmallJoker) [reply builtin/game/tests/test_moveaction.lua:65]: I believe this is the easiest solution. Moving it to C++ side would mean to hijack several callback functions just for this unittest.

File: src/inventorymanager.cpp (hunk 10 above)
Comment (SmallJoker) [reply src/inventorymanager.cpp:451]: how about `std::min<u32>`?

File: src/unittest/test_moveaction.cpp (hunk 12 above)
Comment (SmallJoker) [reply src/unittest/test_moveaction.cpp:62]: `core.get_modpath` (and similar) rely on the existence of `ServerModManager`. When I last checked, the pointer was not `nullptr`, indicating a deeper issue. However, that could also have changed since the last rebase.

File: src/inventorymanager.cpp (hunk 10 above)
Comment (sfan5) [reply src/inventorymanager.cpp:451]: ah, sure.

File: src/inventorymanager.cpp (hunk 11 above)
Comment (SmallJoker) [reply src/inventorymanager.cpp:455]: I'll instead reword the comment above to clarify that the ItemStack might still differ due to allow_* callback execution, even though mods should not use that to modify the inventory.

**Comment by sfan5** [comment]
question: does this also fix #11805?

**Comment by SmallJoker** [comment]
> question: does this also fix #11805?

#11805 is older than the UAF/out-of-bound fixes (#13360 and #13919) and there is afaik yet no reliable way to reproduce that error. @Bastrabun @debagos Would you please be so nice to test whether this PR fixes your odd crashes - or did the crash not occur any more in the past few months (5.9.0-dev builds)?

I somewhat believe that this particular issue is resolved since commit bec9c68bf3d.

**Review by sfan5 — COMMENTED** [review]
Funny behavior:
a

**Comment by SmallJoker** [comment]
@sfan5 This particular effect can be solved with the following change:
```diff
diff --git a/src/gui/guiFormSpecMenu.cpp b/src/gui/guiFormSpecMenu.cpp
index 0c616967c..9390ed20c 100644
--- a/src/gui/guiFormSpecMenu.cpp
+++ b/src/gui/guiFormSpecMenu.cpp
@@ -3891,8 +3891,11 @@ ItemStack GUIFormSpecMenu::verifySelectedItem()
                                        ItemStack stack = list->getItem(m_selected_item->i);
                                        if (!m_selected_swap.empty()) {
                                                if (m_selected_swap.name == stack.name &&
-                                                               m_selected_swap.count == stack.count)
+                                                               m_selected_swap.count == stack.count) {
                                                        m_selected_swap.clear();
+                                               } else {
+                                                       m_selected_swap = stack;
+                                               }
                                        } else {
                                                m_selected_amount = std::min(m_selected_amount, stack.count);
                                        }
```

But I will not include it in this PR because that code is even more cursed that what I already touched.

**Comment by sfan5** [comment]
So it's not a new bug? I don't remember seeing it before.

**Review by sfan5 — APPROVED** [review]
I carefully reviewed the code previous time, did some cursory testing now and hopefully the unit tests cover enough stuff so this should be good to go.

**Comment by SmallJoker** [comment]
> So it's not a new bug? I don't remember seeing it before.

Reproduced in 5.9.0-dev 658bc9fc (master, about 3 days old at the time of writing). That bug already exists independently to this PR.
